File: src/brs/services/impl/AccountServiceImpl.java
package brs.services.impl;

import brs.Account;
import brs.Account.AccountAsset;
import brs.Account.Event;
import brs.Account.RewardRecipientAssignment;
import brs.AssetTransfer;
import brs.Signum;
import brs.Constants;
import brs.crypto.Crypto;
import brs.db.SignumKey;
import brs.db.SignumKey.LinkKeyFactory;
import brs.db.SignumKey.LongKeyFactory;
import brs.db.VersionedBatchEntityTable;
import brs.db.VersionedEntityTable;
import brs.db.store.AccountStore;
import brs.db.store.AssetTransferStore;
import brs.services.AccountService;
import brs.util.CollectionWithIndex;
import brs.util.Convert;
import brs.util.Listener;
import brs.util.Listeners;

import java.util.Arrays;
import java.util.Collection;

import static brs.schema.Tables.ACCOUNT;

public class AccountServiceImpl implements AccountService {

  private final AccountStore accountStore;
  private final VersionedBatchEntityTable<Account> accountTable;
  private final VersionedBatchEntityTable<Account.Balance> accountBalanceTable;
  private final LongKeyFactory<Account> accountSignumKeyFactory;
  private final LongKeyFactory<Account.Balance> accountBalanceSignumKeyFactory;
  private final VersionedEntityTable<AccountAsset> accountAssetTable;
  private final LinkKeyFactory<AccountAsset> accountAssetKeyFactory;
  private final VersionedEntityTable<RewardRecipientAssignment> rewardRecipientAssignmentTable;
  private final LongKeyFactory<RewardRecipientAssignment> rewardRecipientAssignmentKeyFactory;

  private final AssetTransferStore assetTransferStore;

  private final Listeners<Account, Event> listeners = new Listeners<>();
  private final Listeners<AccountAsset, Event> assetListeners = new Listeners<>();

  public AccountServiceImpl(AccountStore accountStore, AssetTransferStore assetTransferStore) {
    this.accountStore = accountStore;
    this.accountTable = accountStore.getAccountTable();
    this.accountBalanceTable = accountStore.getAccountBalanceTable();
    this.accountSignumKeyFactory = accountStore.getAccountKeyFactory();
    this.accountBalanceSignumKeyFactory = accountStore.getAccountBalanceKeyFactory();
    this.assetTransferStore = assetTransferStore;
    this.accountAssetTable = accountStore.getAccountAssetTable();
    this.accountAssetKeyFactory = accountStore.getAccountAssetKeyFactory();
    this.rewardRecipientAssignmentTable = accountStore.getRewardRecipientAssignmentTable();
    this.rewardRecipientAssignmentKeyFactory = accountStore.getRewardRecipientAssignmentKeyFactory();
  }

  @Override
  public boolean addListener(Listener<Account> listener, Event eventType) {
    return listeners.addListener(listener, eventType);
  }

  @Override
  public boolean addAssetListener(Listener<AccountAsset> listener, Event eventType) {
    return assetListeners.addListener(listener, eventType);
  }

  @Override
  public Account getAccount(long id) {
    return id == 0 ? null : accountTable.get(accountSignumKeyFactory.newKey(id));
  }

  @Override
  public Account.Balance getAccountBalance(long id) {
    Account.Balance account = accountBalanceTable.get(accountBalanceSignumKeyFactory.newKey(id));
    if(account == null){
      account = new Account.Balance(id);
    }
    return account;
  }

  @Override
  public Account getNullAccount() {
    return accountTable.get(accountSignumKeyFactory.newKey(0L));
  }

  @Override
  public Account getAccount(long id, int height) {
    return id == 0 ? null : accountTable.get(accountSignumKeyFactory.newKey(id), height);
  }

  @Override
  public Account getAccount(byte[] publicKey) {
    final Account account = accountTable.get(accountSignumKeyFactory.newKey(getId(publicKey)));

    if (account == null) {
      return null;
    }

    if (account.getPublicKey() == null || Arrays.equals(account.getPublicKey(), publicKey)) {
      return account;
    }

    throw new RuntimeException("DUPLICATE KEY for account " + Convert.toUnsignedLong(account.getId())
        + " existing key " + Convert.toHexString(account.getPublicKey()) + " new key " + Convert.toHexString(publicKey));
  }

  @Override
  public CollectionWithIndex<AssetTransfer> getAssetTransfers(long accountId, int from, int to) {
    return new CollectionWithIndex<AssetTransfer>(assetTransferStore.getAccountAssetTransfers(accountId, from, to), from, to);
  }

  @Override
  public CollectionWithIndex<AccountAsset> getAssets(long accountId, int from, int to) {
    return new CollectionWithIndex<AccountAsset>(accountStore.getAssets(from, to, accountId), from, to);
  }

  @Override
  public Collection<RewardRecipientAssignment> getAccountsWithRewardRecipient(Long recipientId) {
    return accountStore.getAccountsWithRewardRecipient(recipientId);
  }

  @Override
  public Collection<Account> getAccountsWithName(String name) {
    return accountTable.getManyBy(ACCOUNT.NAME.equalIgnoreCase(name), 0, -1);
  }

  @Override
  public CollectionWithIndex<Account> getAllAccounts(int from, int to) {
    return new CollectionWithIndex<Account>(accountTable.getAll(from, to), from, to);
  }

  @Override
  public long getAllAccountsBalance() {
    return accountStore.getAllAccountsBalance();
  }

  @Override
  public Account getOrAddAccount(long id) {
    Account account = accountTable.get(accountSignumKeyFactory.newKey(id));
    if (account == null) {
      account = new Account(id);
      accountTable.insert(account);
    }
    return account;
  }

  public static long getId(byte[] publicKey) {
    byte[] publicKeyHash = Crypto.sha256().digest(publicKey);
    return Convert.fullHashToId(publicKeyHash);
  }

  @Override
  public void flushAccountTable() {
    accountTable.finish();
  }

  @Override
  public int getCount() {
    return accountTable.getCount();
  }

  @Override
  public int getBatchedAccountsCount() {
    return accountTable.getBatch().size();
  }

  @Override
  public void addToForgedBalanceNQT(Account account, long amountNQT) {
    if (amountNQT == 0) {
      return;
    }
    Account.Balance accountBalance = getAccountBalance(account.getId());
    accountBalance.setForgedBalanceNqt(Convert.safeAdd(accountBalance.getForgedBalanceNqt(), amountNQT));
    accountBalanceTable.insert(accountBalance);
  }

  @Override
  public void setAccountInfo(Account account, String name, String description) {
    account.setName(Convert.emptyToNull(name.trim()));
    account.setDescription(Convert.emptyToNull(description.trim()));
    accountTable.insert(account);
  }

  @Override
  public void addToAssetBalanceQNT(Account account, long assetId, long quantityQNT) {
    if (quantityQNT == 0) {
      return;
    }
    AccountAsset accountAsset;

    SignumKey newKey = accountAssetKeyFactory.newKey(account.getId(), assetId);
    accountAsset = accountAssetTable.get(newKey);
    long assetBalance = accountAsset == null ? 0 : accountAsset.getQuantityQnt();
    assetBalance = Convert.safeAdd(assetBalance, quantityQNT);
    if (accountAsset == null) {
      accountAsset = new AccountAsset(newKey, account.getId(), assetId, assetBalance, 0);
    } else {
      accountAsset.setQuantityQnt(assetBalance);
    }
    saveAccountAsset(accountAsset);
    listeners.notify(account, Event.ASSET_BALANCE);
    assetListeners.notify(accountAsset, Event.ASSET_BALANCE);
  }

  @Override
  public void addToUnconfirmedAssetBalanceQNT(Account account, long assetId, long quantityQNT) {
    if (quantityQNT == 0) {
      return;
    }
    AccountAsset accountAsset;
    SignumKey newKey = accountAssetKeyFactory.newKey(account.getId(), assetId);
    accountAsset = accountAssetTable.get(newKey);
    long unconfirmedAssetBalance = accountAsset == null ? 0 : accountAsset.getUnconfirmedQuantityQnt();
    unconfirmedAssetBalance = Convert.safeAdd(unconfirmedAssetBalance, quantityQNT);
    if (accountAsset == null) {
      accountAsset = new AccountAsset(newKey, account.getId(), assetId, 0, unconfirmedAssetBalance);
    } else {
      accountAsset.setUnconfirmedQuantityQnt(unconfirmedAssetBalance);
    }
    saveAccountAsset(accountAsset);
    listeners.notify(account, Event.UNCONFIRMED_ASSET_BALANCE);
    assetListeners.notify(accountAsset, Event.UNCONFIRMED_ASSET_BALANCE);
  }

  @Override
  public void addToAssetAndUnconfirmedAssetBalanceQNT(Account account, long assetId, long quantityQNT) {
    if (quantityQNT == 0) {
      return;
    }
    AccountAsset accountAsset;
    SignumKey newKey = accountAssetKeyFactory.newKey(account.getId(), assetId);
    accountAsset = accountAssetTable.get(newKey);
    long assetBalance = accountAsset == null ? 0 : accountAsset.getQuantityQnt();
    assetBalance = Convert.safeAdd(assetBalance, quantityQNT);
    long unconfirmedAssetBalance = accountAsset == null ? 0 : accountAsset.getUnconfirmedQuantityQnt();
    unconfirmedAssetBalance = Convert.safeAdd(unconfirmedAssetBalance, quantityQNT);
    if (accountAsset == null) {
      accountAsset = new AccountAsset(newKey, account.getId(), assetId, assetBalance, unconfirmedAssetBalance);
    } else {
      accountAsset.setQuantityQnt(assetBalance);
      accountAsset.setUnconfirmedQuantityQnt(unconfirmedAssetBalance);
    }
    saveAccountAsset(accountAsset);
    listeners.notify(account, Event.ASSET_BALANCE);
    listeners.notify(account, Event.UNCONFIRMED_ASSET_BALANCE);
    assetListeners.notify(accountAsset, Event.ASSET_BALANCE);
    assetListeners.notify(accountAsset, Event.UNCONFIRMED_ASSET_BALANCE);
  }

  @Override
  public void addToBalanceNQT(Account account, long amountNQT) {
    if (amountNQT == 0) {
      return;
    }
    Account.Balance accountBalance = getAccountBalance(account.getId());

    accountBalance.setBalanceNqt(Convert.safeAdd(accountBalance.getBalanceNqt(), amountNQT));
    accountBalance.checkBalance();
    accountBalanceTable.insert(accountBalance);
    listeners.notify(account, Event.BALANCE);
  }

  @Override
  public void addToUnconfirmedBalanceNQT(Account account, long amountNQT) {
    if (amountNQT == 0) {
      return;
    }
    Account.Balance accountBalance = getAccountBalance(account.getId());

    accountBalance.setUnconfirmedBalanceNqt(Convert.safeAdd(accountBalance.getUnconfirmedBalanceNqt(), amountNQT));
    accountBalance.checkBalance();
    accountBalanceTable.insert(accountBalance);
    listeners.notify(account, Event.UNCONFIRMED_BALANCE);
  }

  @Override
  public void addToBalanceAndUnconfirmedBalanceNQT(Account account, long amountNQT) {
    if (amountNQT == 0) {
      return;
    }
    Account.Balance accountBalance = getAccountBalance(account.getId());

    accountBalance.setBalanceNqt(Convert.safeAdd(accountBalance.getBalanceNqt(), amountNQT));
    accountBalance.setUnconfirmedBalanceNqt(Convert.safeAdd(accountBalance.getUnconfirmedBalanceNqt(), amountNQT));
    accountBalance.checkBalance();
    accountBalanceTable.insert(accountBalance);
    listeners.notify(account, Event.BALANCE);
    listeners.notify(account, Event.UNCONFIRMED_BALANCE);
  }

  @Override
  public RewardRecipientAssignment getRewardRecipientAssignment(Account account) {
    return getRewardRecipientAssignment(account.getId());
  }

  private RewardRecipientAssignment getRewardRecipientAssignment(Long id) {
    return rewardRecipientAssignmentTable.get(rewardRecipientAssignmentKeyFactory.newKey(id));
  }

  @Override
  public void setRewardRecipientAssignment(Account account, Long recipient) {
    int currentHeight = Signum.getBlockchain().getLastBlock().getHeight();
    RewardRecipientAssignment assignment = getRewardRecipientAssignment(account.getId());
    if (assignment == null) {
      SignumKey signumKey = rewardRecipientAssignmentKeyFactory.newKey(account.getId());
      assignment = new RewardRecipientAssignment(account.getId(), account.getId(), recipient, (int) (currentHeight + Constants.SIGNA_REWARD_RECIPIENT_ASSIGNMENT_WAIT_TIME), signumKey);
    } else {
      assignment.setRecipient(recipient, (int) (currentHeight + Constants.SIGNA_REWARD_RECIPIENT_ASSIGNMENT_WAIT_TIME));
    }
    rewardRecipientAssignmentTable.insert(assignment);
  }

  @Override
  public long getUnconfirmedAssetBalanceQNT(Account account, long assetId) {
    SignumKey newKey = Signum.getStores().getAccountStore().getAccountAssetKeyFactory().newKey(account.getId(), assetId);
    AccountAsset accountAsset = accountAssetTable.get(newKey);
    return accountAsset == null ? 0 : accountAsset.getUnconfirmedQuantityQnt();
  }


  private void saveAccountAsset(AccountAsset accountAsset) {
    accountAsset.checkBalance();
    if (accountAsset.getQuantityQnt() > 0 || accountAsset.getUnconfirmedQuantityQnt() > 0) {
      accountAssetTable.insert(accountAsset);
    } else {
      accountAssetTable.delete(accountAsset);
    }
  }
}

File: src/main/kotlin/brs/services/impl/AccountServiceImpl.kt
package brs.services.impl

import brs.db.AccountStore
import brs.db.AssetTransferStore
import brs.db.BurstKey.LinkKeyFactory
import brs.db.BurstKey.LongKeyFactory
import brs.db.VersionedBatchEntityTable
import brs.db.VersionedEntityTable
import brs.entity.Account
import brs.entity.Account.*
import brs.entity.AssetTransfer
import brs.entity.DependencyProvider
import brs.objects.Constants
import brs.schema.Tables.ACCOUNT
import brs.services.AccountService
import brs.util.Listeners
import brs.util.convert.fullHashToId
import brs.util.convert.safeAdd
import brs.util.convert.toHexString
import brs.util.convert.toUnsignedString
import brs.util.crypto.Crypto
import java.util.*

class AccountServiceImpl(private val dp: DependencyProvider) : AccountService {
    private val accountStore: AccountStore
    private val accountTable: VersionedBatchEntityTable<Account>
    private val accountBurstKeyFactory: LongKeyFactory<Account>
    private val accountAssetTable: VersionedEntityTable<AccountAsset>
    private val accountAssetKeyFactory: LinkKeyFactory<AccountAsset>
    private val rewardRecipientAssignmentTable: VersionedEntityTable<RewardRecipientAssignment>
    private val rewardRecipientAssignmentKeyFactory: LongKeyFactory<RewardRecipientAssignment>

    private val assetTransferStore: AssetTransferStore

    private val listeners = Listeners<Account, Event>()
    private val assetListeners = Listeners<AccountAsset, Event>()

    override val count get() = accountTable.count

    init { // TODO don't hold references to dependency instances
        val accountStore = dp.accountStore
        this.accountStore = accountStore
        this.accountTable = accountStore.accountTable
        this.accountBurstKeyFactory = accountStore.accountKeyFactory
        this.assetTransferStore = dp.assetTransferStore
        this.accountAssetTable = accountStore.accountAssetTable
        this.accountAssetKeyFactory = accountStore.accountAssetKeyFactory
        this.rewardRecipientAssignmentTable = accountStore.rewardRecipientAssignmentTable
        this.rewardRecipientAssignmentKeyFactory = accountStore.rewardRecipientAssignmentKeyFactory
    }

    override fun addListener(eventType: Event, listener: (Account) -> Unit) {
        listeners.addListener(eventType, listener)
    }

    override fun addAssetListener(
        eventType: Event,
        listener: (AccountAsset) -> Unit
    ) {
        assetListeners.addListener(eventType, listener)
    }

    override fun getAccount(id: Long): Account? {
        return if (id == 0L) null else accountTable[accountBurstKeyFactory.newKey(id)]
    }

    override fun getAccount(id: Long, height: Int): Account? {
        return if (id == 0L) null else accountTable[accountBurstKeyFactory.newKey(id), height]
    }

    override fun getAccount(publicKey: ByteArray): Account? {
        val account = accountTable[accountBurstKeyFactory.newKey(getId(publicKey))] ?: return null

        if (account.publicKey == null || Arrays.equals(account.publicKey, publicKey)) {
            return account
        }

        throw RuntimeException("DUPLICATE KEY for account " + account.id.toUnsignedString() + " existing key " + account.publicKey.toHexString() + " new key " + publicKey.toHexString())
    }

    override fun getAssetTransfers(accountId: Long, from: Int, to: Int): Collection<AssetTransfer> {
        return assetTransferStore.getAccountAssetTransfers(accountId, from, to)
    }

    override fun getAssets(accountId: Long, from: Int, to: Int): Collection<AccountAsset> {
        return accountStore.getAssets(from, to, accountId)
    }

    override fun getAccountsWithRewardRecipient(recipientId: Long?): Collection<RewardRecipientAssignment> {
        return accountStore.getAccountsWithRewardRecipient(recipientId)
    }

    override fun getAccountsWithName(name: String): Collection<Account> {
        return accountTable.getManyBy(ACCOUNT.NAME.equalIgnoreCase(name), 0, -1)
    }

    override fun getAllAccounts(from: Int, to: Int): Collection<Account> {
        return accountTable.getAll(from, to)
    }

    override fun getOrAddAccount(id: Long): Account {
        var account: Account? = accountTable[accountBurstKeyFactory.newKey(id)]
        if (account == null) {
            account = Account(dp, id)
            accountTable.insert(account)
        }
        return account
    }

    override fun flushAccountTable() {
        accountTable.finish()
    }

    override fun addToForgedBalancePlanck(account: Account, amountPlanck: Long) {
        if (amountPlanck == 0L) {
            return
        }
        account.forgedBalancePlanck = account.forgedBalancePlanck.safeAdd(amountPlanck)
        accountTable.insert(account)
    }

    override fun setAccountInfo(account: Account, name: String, description: String) {
        account.name = name.trim { it <= ' ' }
        account.description = description.trim { it <= ' ' }
        accountTable.insert(account)
    }

    override fun addToAssetBalanceQuantity(account: Account, assetId: Long, quantity: Long) {
        if (quantity == 0L) {
            return
        }
        var accountAsset: AccountAsset?

        val newKey = accountAssetKeyFactory.newKey(account.id, assetId)
        accountAsset = accountAssetTable[newKey]
        var assetBalance = accountAsset?.quantity ?: 0
        assetBalance = assetBalance.safeAdd(quantity)
        if (accountAsset == null) {
            accountAsset = AccountAsset(newKey, account.id, assetId, assetBalance, 0)
        } else {
            accountAsset.quantity = assetBalance
        }
        saveAccountAsset(accountAsset)
        listeners.accept(Event.ASSET_BALANCE, account)
        assetListeners.accept(Event.ASSET_BALANCE, accountAsset)
    }

    override fun addToUnconfirmedAssetBalanceQuantity(account: Account, assetId: Long, quantity: Long) {
        if (quantity == 0L) {
            return
        }
        var accountAsset: AccountAsset?
        val newKey = accountAssetKeyFactory.newKey(account.id, assetId)
        accountAsset = accountAssetTable[newKey]
        var unconfirmedAssetBalance = accountAsset?.unconfirmedQuantity ?: 0
        unconfirmedAssetBalance = unconfirmedAssetBalance.safeAdd(quantity)
        if (accountAsset == null) {
            accountAsset = AccountAsset(newKey, account.id, assetId, 0, unconfirmedAssetBalance)
        } else {
            accountAsset.unconfirmedQuantity = unconfirmedAssetBalance
        }
        saveAccountAsset(accountAsset)
        listeners.accept(Event.UNCONFIRMED_ASSET_BALANCE, account)
        assetListeners.accept(Event.UNCONFIRMED_ASSET_BALANCE, accountAsset)
    }

    override fun addToAssetAndUnconfirmedAssetBalanceQuantity(account: Account, assetId: Long, quantity: Long) {
        if (quantity == 0L) {
            return
        }
        var accountAsset: AccountAsset?
        val newKey = accountAssetKeyFactory.newKey(account.id, assetId)
        accountAsset = accountAssetTable[newKey]
        var assetBalance = accountAsset?.quantity ?: 0
        assetBalance = assetBalance.safeAdd(quantity)
        var unconfirmedAssetBalance = accountAsset?.unconfirmedQuantity ?: 0
        unconfirmedAssetBalance = unconfirmedAssetBalance.safeAdd(quantity)
        if (accountAsset == null) {
            accountAsset = AccountAsset(newKey, account.id, assetId, assetBalance, unconfirmedAssetBalance)
        } else {
            accountAsset.quantity = assetBalance
            accountAsset.unconfirmedQuantity = unconfirmedAssetBalance
        }
        saveAccountAsset(accountAsset)
        listeners.accept(Event.ASSET_BALANCE, account)
        listeners.accept(Event.UNCONFIRMED_ASSET_BALANCE, account)
        assetListeners.accept(Event.ASSET_BALANCE, accountAsset)
        assetListeners.accept(Event.UNCONFIRMED_ASSET_BALANCE, accountAsset)
    }

    override fun addToBalancePlanck(account: Account, amountPlanck: Long) {
        if (amountPlanck == 0L) {
            return
        }
        account.balancePlanck = account.balancePlanck.safeAdd(amountPlanck)
        account.checkBalance()
        accountTable.insert(account)
        listeners.accept(Event.BALANCE, account)
    }

    override fun addToUnconfirmedBalancePlanck(account: Account, amountPlanck: Long) {
        if (amountPlanck == 0L) {
            return
        }
        account.unconfirmedBalancePlanck = account.unconfirmedBalancePlanck.safeAdd(amountPlanck)
        account.checkBalance()
        accountTable.insert(account)
        listeners.accept(Event.UNCONFIRMED_BALANCE, account)
    }

    override fun addToBalanceAndUnconfirmedBalancePlanck(account: Account, amountPlanck: Long) {
        if (amountPlanck == 0L) {
            return
        }
        account.balancePlanck = account.balancePlanck.safeAdd(amountPlanck)
        account.unconfirmedBalancePlanck = account.unconfirmedBalancePlanck.safeAdd(amountPlanck)
        account.checkBalance()
        accountTable.insert(account)
        listeners.accept(Event.BALANCE, account)
        listeners.accept(Event.UNCONFIRMED_BALANCE, account)
    }

    override fun getRewardRecipientAssignment(account: Account): RewardRecipientAssignment? {
        return rewardRecipientAssignmentTable[rewardRecipientAssignmentKeyFactory.newKey(account.id)]
    }

    override fun setRewardRecipientAssignment(account: Account, recipient: Long) {
        val currentHeight = dp.blockchainService.height
        var assignment = getRewardRecipientAssignment(account)
        if (assignment == null) {
            val burstKey = rewardRecipientAssignmentKeyFactory.newKey(account.id)
            assignment = RewardRecipientAssignment(
                account.id,
                account.id,
                recipient,
                (currentHeight + Constants.BURST_REWARD_RECIPIENT_ASSIGNMENT_WAIT_TIME).toInt(),
                burstKey
            )
        } else {
            assignment.setRecipient(
                recipient,
                (currentHeight + Constants.BURST_REWARD_RECIPIENT_ASSIGNMENT_WAIT_TIME).toInt()
            )
        }
        rewardRecipientAssignmentTable.insert(assignment)
    }

    override fun getUnconfirmedAssetBalanceQuantity(account: Account, assetId: Long): Long {
        val newKey = accountAssetKeyFactory.newKey(account.id, assetId)
        val accountAsset = accountAssetTable[newKey]
        return accountAsset?.unconfirmedQuantity ?: 0
    }


    private fun saveAccountAsset(accountAsset: AccountAsset) {
        accountAsset.checkBalance()
        if (accountAsset.quantity > 0 || accountAsset.unconfirmedQuantity > 0) {
            accountAssetTable.insert(accountAsset)
        } else {
            accountAssetTable.delete(accountAsset)
        }
    }

    companion object {

        fun getId(publicKey: ByteArray): Long {
            val publicKeyHash = Crypto.sha256().digest(publicKey)
            return publicKeyHash.fullHashToId()
        }
    }
}
